File: Haskell/04_functions_recursion_io_jump.hs
-- 任务：函数及其递归；文件和控制台 IO；跳转。
-- 说明：Haskell 没有 goto/longjmp；非局部跳转通常用异常实现。
-- 运行：runghc 04_04_functions_recursion_io_jump.hs

import Control.Exception as E
import Text.Printf (printf)


-- ---------- 函数及其递归 ----------

factorial :: Integer -> Integer
factorial n = if n <= 1 then 1 else n * factorial (n - 1)

ackermann :: Integer -> Integer -> Integer
ackermann m n
    | m == 0    = n + 1
    | n == 0    = ackermann (m - 1) 1
    | otherwise = ackermann (m - 1) (ackermann m (n - 1))

isEven, isOdd :: Int -> Bool
isEven 0 = True
isEven n = isOdd (n - 1)
isOdd 0 = False
isOdd n = isEven (n - 1)          -- 互递归


-- ---------- 文件和控制台 IO ----------

ioDemo :: IO ()
ioDemo = do
    let path = "04_04_functions_recursion_io_jump_demo.txt"
    writeFile path "line one\nline two\n"
    content <- readFile path
    putStrLn ("file content: " ++ show content)

    putStr "console input (one line): "
    line <- E.catch getLine (\(_ :: E.IOException) -> return "(EOF/no input)")
    putStrLn ("you typed: " ++ line)


-- ---------- 跳转 ----------

-- 纯函数中的提前返回：用 Either 表示成功 / 失败。
earlyReturn :: Int -> Either String Int
earlyReturn x
    | x < 0     = Left "negative input"
    | otherwise = Right (x * 2)

-- IO 中的非局部跳转：抛出并捕获异常。
jumpDemo :: IO ()
jumpDemo = do
    print (earlyReturn (-3) :: Either String Int)

    E.catch
        (do
            putStrLn "before jump"
            E.throwIO (E.ErrorCall "jump out")
            putStrLn "never printed")
        (\(E.ErrorCall msg) -> putStrLn ("caught: " ++ msg))


main :: IO ()
main = do
    printf "factorial(10)=%d\n" (factorial 10)
    printf "ackermann(2,3)=%d\n" (ackermann 2 3)
    printf "isEven(10)=%s, isOdd(10)=%s\n" (show (isEven 10)) (show (isOdd 10))
    putStrLn ""
    ioDemo
    putStrLn ""
    jumpDemo

File: C/04_functions_recursion_io_jump.c
/*
 * 任务：函数及其递归；文件和控制台 IO；跳转。
 * 编译：gcc -std=c11 -O2 -Wall -Wextra 04_04_functions_recursion_io_jump.c
 */

#include <stdio.h>
#include <stdlib.h>
#include <setjmp.h>


/* ---------- 函数及其递归 ---------- */

static unsigned long long factorial(int n) {
    return n <= 1 ? 1ULL : (unsigned long long)n * factorial(n - 1);
}

static int ackermann(int m, int n) {
    if (m == 0) return n + 1;
    if (n == 0) return ackermann(m - 1, 1);
    return ackermann(m - 1, ackermann(m, n - 1));
}

/* 互递归：is_even 与 is_odd 互相调用。 */
static int is_odd(int n);
static int is_even(int n) { return n == 0 ? 1 : is_odd(n - 1); }
static int is_odd(int n)  { return n == 0 ? 0 : is_even(n - 1); }


/* ---------- 文件和控制台 IO ---------- */

static void io_demo(void) {
    const char *path = "04_04_functions_recursion_io_jump_demo.txt";
    FILE *fp = fopen(path, "w");
    if (!fp) { perror("fopen(w)"); return; }
    fprintf(fp, "line one\nline two\n");
    fclose(fp);

    fp = fopen(path, "r");
    if (!fp) { perror("fopen(r)"); return; }
    char buf[100];
    printf("file content: ");
    while (fgets(buf, sizeof buf, fp)) printf("[%s]", buf);
    fclose(fp);

    printf("console input (one line): ");
    fflush(stdout);
    if (fgets(buf, sizeof buf, stdin))
        printf("you typed: %s", buf);
    else
        printf("(EOF)\n");
}


/* ---------- 跳转：goto 与 setjmp/longjmp ---------- */

static jmp_buf jump_env;

static void may_fail(int x) {
    if (x < 0) longjmp(jump_env, 1);   /* 非局部跳转 */
    printf("may_fail: x=%d\n", x);
}

static void jump_demo(void) {
    int i = 0;
    printf("goto loop: ");
again:
    printf("%d ", i);
    ++i;
    if (i < 5) goto again;
    printf("\n");

    if (setjmp(jump_env) == 0) {
        printf("before longjmp\n");
        may_fail(-1);
        printf("never printed\n");
    } else {
        printf("after longjmp: caught in calling function\n");
    }
}


int main(void) {
    printf("factorial(10)=%llu\n", factorial(10));
    printf("ackermann(2,3)=%d\n", ackermann(2, 3));
    printf("is_even(10)=%d, is_odd(10)=%d\n", is_even(10), is_odd(10));
    printf("\n");
    io_demo();
    printf("\n");
    jump_demo();
    return 0;
}